aesara/tensor/shape.py

```diff
@@ -396,136 +376,153 @@ class SpecifyShape(COp):
+        for i, (xts, s) in enumerate(zip(x.type.shape, shape)):
+            if xts is not None:
+                type_shape[i] = xts
+            else:
+                try:
+                    type_s = get_scalar_constant_value(s)
+                    if type_s is not None:
+                        type_shape[i] = int(type_s)
+                except NotScalarConstantError:
+                    pass
+
+        out_var = x.type.clone(shape=type_shape)()
 
-        in_shape = at.as_tensor_variable(shape, ndim=1)
-        return Apply(self, [x, in_shape], [out_var])
+        return Apply(self, [x, *shape], [out_var])
 
     def perform(self, node, inp, out_):
-        x, shape = inp
+        x, *shape = inp
         (out,) = out_
         ndim = len(shape)
         if x.ndim != ndim:
             raise AssertionError(
                 f"SpecifyShape: Got {x.ndim} dimensions (shape {x.shape}), expected {ndim} dimensions with shape {tuple(shape)}."
             )
-        if x.shape != tuple(shape):
+        if not all(xs == s for xs, s in zip(x.shape, shape) if s is not None):
             raise AssertionError(
-                f"SpecifyShape: Got shape {x.shape}, expected {tuple(shape)}."
+                f"SpecifyShape: Got shape {x.shape}, expected {tuple(int(s) if s is not None else None for s in shape)}."
             )
         out[0] = x
 
     def infer_shape(self, fgraph, node, shapes):
-        xshape, sshape = shapes
+        xshape, *_ = shapes
+        shape = node.inputs[1:]
         new_shape = []
         for dim in range(node.inputs[0].type.ndim):
             try:
-                s = at.get_scalar_constant_value(node.inputs[1][dim])
-                s = at.as_tensor_variable(s)
-                new_shape.append(s)
+                s = at.get_scalar_constant_value(shape[dim])
+                if s is None:
+                    s = xshape[dim]
             except NotScalarConstantError:
-                new_shape.append(node.inputs[1][dim])
+                s = xshape[dim]
+            new_shape.append(at.as_tensor_variable(s))
 
         assert len(new_shape) == len(xshape)
         return [new_shape]
 
     def connection_pattern(self, node):
-        return [[True], [False]]
+        return [[True], *[[False]] * len(node.inputs[1:])]
 
     def grad(self, inp, grads):
-        x, s = inp
+        x, *shape = inp
         (gz,) = grads
         # Should I set an SpecifyShape on gz? I think so
         # But I don't do it now as we need to make an optimization
         # to remove that op from the graph to don't block other optimization
         # Should I do an optimizer that will remove the SpecifyShape?
         # I think Yes
-        return [gz, aesara.gradient.DisconnectedType()()]
-        return [specify_shape(gz, s), aesara.gradient.DisconnectedType()()]
+        # return [specify_shape(gz, s)] + [aesara.gradient.DisconnectedType()()] * len(shape)
+        return [gz] + [aesara.gradient.DisconnectedType()()] * len(shape)
```

Comment:
Check: Do I need a unique DisconnectedType per input or is the multiplication fine

Reply:
I'm not entirely sure how the `Variable` instances of these are used in every scenario, but I do see that distinct ones are created for each distinct term to which they correspond, so it would be safer to create unique ones.